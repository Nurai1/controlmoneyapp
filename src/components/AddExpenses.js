import React from 'react';
import { connect } from 'react-redux';

import { addExpenses } from '../actions';
import { isPositiveNumber } from '../utilities';

const AddExpenses = ({
  generalSum,
  isCurrentSumPositive,
  dispatch,
}) => {
  let expenseName = '';
  let expenseValue = '';

  const addExpenseToList = () => {
    if (expenseName.value.trim() === '' || expenseValue.value.trim() === ''
        || !isPositiveNumber(expenseValue.value)) { return; }

    dispatch(addExpenses(Number(expenseValue.value), expenseName.value));

    expenseName.value = '';
    expenseValue.value = '';
  };

  if (isPositiveNumber(generalSum)) {
    return (
      <div className="addExpense">
        <p className="addExpense__info">
          Введите вид товара, на который вы собираетесь тратить деньги и сколько
          вы планируете на него потратить.
        </p>
        <label>
          Вид товара:
          <br />
          <input ref={(input) => { expenseName = input; }} type="text" />
          <br />
        </label>
        <label>
          Планируемые расходы:
          <br />
          <input ref={(input) => { expenseValue = input; }} type="text" />
          <br />
        </label>
        <button type="button" onClick={addExpenseToList}>Добавить расходы</button>
        <p>
          {isCurrentSumPositive ? ''
            : 'Не хватает средств, чтобы добавить категорию'}
        </p>
      </div>
    );
  }

  return '';
};

const mapStateToProps = (state) => ({
  generalSum: state.generalSum,
  isCurrentSumPositive: state.isCurrentSumPositive,
});

const AddExpensesContainer = connect(
  mapStateToProps,
)(AddExpenses);

export default AddExpensesContainer;
