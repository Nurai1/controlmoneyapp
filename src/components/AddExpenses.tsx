import React from 'react';
import { connect } from 'react-redux';

import { addExpenses } from '../store/actions';
import { isPositiveNumber } from '../utils/isPositiveNum';

import {
    AppState,
    addExpenseAction,
    ExpensePayload
} from '../store/types';

interface AddExpensesProps {
    generalSum: number,
        isCurrentSumPositive: boolean,
        addExpenses: (payload: ExpensePayload) => addExpenseAction,
}

const AddExpenses: React.FC<AddExpensesProps> = ({
  generalSum,
  isCurrentSumPositive,
                                                     addExpenses,
}) => {
  let expenseName: HTMLInputElement | null = null;
  let expenseValue: HTMLInputElement | null = null;

  const addExpenseToList = () => {
    if ((expenseName as HTMLInputElement).value.trim() === '' || (expenseValue as HTMLInputElement).value.trim() === ''
        || !isPositiveNumber((expenseValue as HTMLInputElement).value)) { return; }

    addExpenses({ value: Number((expenseValue as HTMLInputElement).value), name: (expenseName as HTMLInputElement).value });

      (expenseName as HTMLInputElement).value = '';
      (expenseValue as HTMLInputElement).value = '';
  };

  if (isPositiveNumber(generalSum)) {
    return (
      <div className="addExpense">
        <p className="addExpense__info">
          Введите вид товара, на который вы собираетесь тратить деньги и сколько
          вы планируете на него потратить.
        </p>
        <label>
          Вид товара:
          <br />
          <input ref={(input) => { expenseName = input; }} type="text" />
          <br />
        </label>
        <label>
          Планируемые расходы:
          <br />
          <input ref={(input) => { expenseValue = input; }} type="text" />
          <br />
        </label>
        <button type="button" onClick={addExpenseToList}>Добавить расходы</button>
        <p>
          {isCurrentSumPositive ? ''
            : 'Не хватает средств, чтобы добавить категорию'}
        </p>
      </div>
    );
  }

  return null;
};

const mapStateToProps = (state: AppState) => ({
  generalSum: state.generalSum,
  isCurrentSumPositive: state.isCurrentSumPositive,
});

const mapDispatchToProps = {
    addExpenses
};

const AddExpensesContainer = connect(
  mapStateToProps,
    mapDispatchToProps
)(AddExpenses);

export default AddExpensesContainer;
