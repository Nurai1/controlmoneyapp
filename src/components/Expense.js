// i think there is no need 
