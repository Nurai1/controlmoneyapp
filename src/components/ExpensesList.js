import React from 'react';
import { connect } from 'react-redux';
import { NavLink } from 'react-router-dom';

export const ExpensesList = ({ expensesList }) => {
  if (expensesList !== undefined && expensesList.length !== 0) {
    return (
      <ul className="expensesList">
        {expensesList.map((expense, idx) => (
          <li key={idx}>
            <NavLink
              exact
              to={`/${expense.id}`}
            >
              На категорию &quot;
              {expense.name}
              &quot; осталось
              {' '}
              {expense.value}
              {' '}
              рублей.
            </NavLink>
          </li>
        ))}
      </ul>
    );
  }

  return '';
};

const mapStateToProps = (state) => ({
  expensesList: state.expenses,
});

const ExpensesListContainer = connect(
  mapStateToProps,
)(ExpensesList);

export default ExpensesListContainer;
