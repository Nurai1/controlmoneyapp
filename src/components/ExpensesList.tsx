import React from 'react';
import { connect } from 'react-redux';
import { NavLink } from 'react-router-dom';

import {
  AppState,
  Expense
} from '../store/types';

interface ExpensesListProps {
  expensesList: Expense[]
}

export const ExpensesList: React.FC<ExpensesListProps> = ({ expensesList }) => {
  if (expensesList !== undefined && expensesList.length !== 0) {
    return (
      <ul className="expensesList">
        {expensesList.map((expense, idx) => (
          <li key={idx}>
            <NavLink
                exact
                to={`/${expense.id}`}
            >
              На категорию &quot;
              {expense.name}
              &quot; осталось
              {' '}
              {expense.value}
              {' '}
              рублей.
            </NavLink>
          </li>
        ))}
      </ul>
    );
  }

  return null;
};

const mapStateToProps = (state: AppState) => ({
  expensesList: state.expenses,
});

const ExpensesListContainer = connect(
  mapStateToProps,
)(ExpensesList);

export default ExpensesListContainer;
