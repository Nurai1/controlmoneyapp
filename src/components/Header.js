import React from 'react';
import { connect } from 'react-redux'

import { addGeneralSum } from '../actions';

let Header = ({ generalSum, currentSum, dispatch }) => {
  let sumInput ='';

  const getGeneralSum = () => {
    dispatch(addGeneralSum(Number(sumInput.value)))
  }

  if (isNaN(generalSum) && generalSum!==undefined){
    return (
        <div>
          <h2>Добро пожаловать в Control Money App.</h2>
          <h3 style={{color: "red"}}>Пожалуйста, введите корректное число.</h3>
          <label htmlFor="generalSum">Введите сумму, которой вы распологаете.</label>
          <input ref={(input)=>{sumInput=input}} id="generalSum" type="text"/>
          <button onClick={getGeneralSum}>Начать учет</button>
        </div>
    )
  }

  if (typeof generalSum === "number"){
    return (
      <div>
        <h2>Добро пожаловать в Control Money App.</h2>
        <p>Ваша общая сумма в наличии: {generalSum} рублей.</p>
        <p>Ваша сумма в остатке после вычета категорий: {currentSum} рублей.</p>
      </div>
    )
  }
  return(
    <div>
      <h2>Добро пожаловать в Control Money App.</h2>
      <label htmlFor="generalSum">Введите сумму, которой вы распологаете.</label>
      <input ref={(input)=>{sumInput=input}} id="generalSum" type="text"/>
      <button onClick={getGeneralSum}>Начать учет</button>
    </div>
  )
}

const mapStateToProps = (state) => {
  return {
    generalSum: state.generalSum,
    currentSum: state.currentSum
  }
}

Header = connect(
  mapStateToProps
)(Header)

export default Header
