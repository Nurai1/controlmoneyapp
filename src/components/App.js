import React from 'react';

import Topbar from './Topbar';
import AddExpenses from './AddExpenses';
import ExpensesList from './ExpensesList';

function App() {
  return (
    <div className="App wrapper">
      <Topbar />
      <AddExpenses />
      <ExpensesList />
    </div>
  );
}

export default App;
