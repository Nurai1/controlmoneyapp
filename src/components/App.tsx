import React from 'react';

import Topbar from './Topbar';
import AddExpenses from './AddExpenses';
import ExpensesList from './ExpensesList';

function App(): JSX.Element {
  return (
    <div className="App wrapper">
      <Topbar />
      <AddExpenses />
      <ExpensesList />
    </div>
  );
}

export default App;
