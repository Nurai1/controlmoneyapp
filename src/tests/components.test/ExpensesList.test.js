import React from 'react';
import { render, unmountComponentAtNode } from 'react-dom';
import { act } from 'react-dom/test-utils';
import pretty from 'pretty';
import { BrowserRouter as Router, Route } from 'react-router-dom';

import { ExpensesList } from '../../components/ExpensesList';

let container = null;
beforeEach(() => {
  container = document.createElement("div");
  document.body.appendChild(container);
});

afterEach(() => {
  unmountComponentAtNode(container);
  container.remove();
  container = null;
});

const expensesList = [{
  id: 1,
  value: 2000,
  name: 'food',
  purchases: [],
}];

it('renders expenses list', () => {
  act(() => {
    render(<ExpensesList />, container);
  });

  expect(pretty(container.innerHTML)).toMatchSnapshot();

  act(() => {
    render(<Router><ExpensesList expensesList={expensesList} /></Router>, container);
  });

  expect(pretty(container.innerHTML)).toMatchSnapshot();
});
