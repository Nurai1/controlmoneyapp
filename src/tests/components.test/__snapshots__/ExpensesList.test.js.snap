// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders expenses list 1`] = `""`;

exports[`renders expenses list 2`] = `
"<ul class=\\"expensesList\\">
  <li><a href=\\"/1\\">На категорию \\"food\\" осталось 2000 рублей.</a></li>
</ul>"
`;
